---
title: "Comparisons of Differential Expression Tools on Different Data Sources"
output: html_notebook
---

# Setup
```{r}
library(biomaRt)
library(DESeq2)
library(sleuth)
library(tximport)
library(jsonlite)
library(tidyverse)
```

## Setup sample details
```{r}
# Load in samples and point to kallisto path
samples <- read_tsv("../work/input/sample_manifest.tsv")
samples$kallisto_path <- file.path("..","work","intermediate","quant",samples$SRR)
```

```{r}
# Find total reads to compute percent mapped
samples$n_reads <- apply(samples,1,function(x)read_json(file.path(x["kallisto_path"],"run_info.json"))$n_processed)
```

## Transcript <-> Gene Mappings
```{r}
mart <- biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL",
  dataset = "hsapiens_gene_ensembl",
  host = 'grch37.ensembl.org')
t2g <- biomaRt::getBM(attributes = c("ensembl_transcript_id_version", "ensembl_gene_id",
    "external_gene_name"), mart = mart)
```

# Sleuth

```{r}
# Initial sleuth
s_obj <- sleuth_prep(
  sample_to_covariates = rename(samples,path=kallisto_path,sample=SRR)
)
```
```{r}
sleuth_est_mapped <- s_obj$obs_raw %>%
  group_by(sample) %>%
  summarise(kallisto_mapped_reads=sum(est_counts))
sleuth_samples <- samples %>% 
  left_join(sleuth_est_mapped,by=c("SRR"="sample")) %>%
  mutate(percent_mapped=kallisto_mapped_reads/n_reads) 
```

```{r}
samples %>% 
left_join(sleuth_est_mapped,by=c("SRR"="sample")) %>%
     mutate(percent_mapped=kallisto_mapped_reads/n_reads) %>% ggplot(aes(x=percent_mapped)) + geom_histogram(bins=5)
#sleuth_samples <- sleuth_samples %>%
#  filter(percent_mapped >= 0.8)
```


```{r include=FALSE}
# Setup sleuth object
s_obj <- sleuth_prep(
  sample_to_covariates = rename(sleuth_samples,path=kallisto_path,sample=SRR),
  full_model = ~Response,
  transform_fun_counts=function(x)log2(x+1),
  target_mapping = rename(t2g,target_id=ensembl_transcript_id_version),
  aggregation_column = "external_gene_name",
  gene_mode=TRUE
)
```

## PCA 

### Sleuth default
```{r}
plot_pca(s_obj,units="scaled_reads_per_base",color_by = "Response") + geom_label(aes(label=sample),size=3)
```

### Standardized PCA
```{r}
sleuth_rna_norm <- s_obj$obs_norm_filt %>% 
  select(sample,target_id,scaled_reads_per_base) %>%
  spread(key=target_id,value=scaled_reads_per_base) %>%
  column_to_rownames("sample")
#sleuth_rna_norm_std <- scale(sleuth_rna_norm)
sleuth_std_pca <- prcomp(sleuth_rna_norm,center=FALSE,scale.=FALSE)
sleuth_std_pca$x[,1:2] %>%
  as.data.frame() %>%
  rownames_to_column("SRR") %>%
  left_join(samples,by="SRR") %>%
  ggplot(aes(x=PC1,y=PC2,color=Response,label=SRR)) +
    geom_point() +
    geom_label(size=2)
```

## Diffex

### LRT
```{r}
s_obj <- s_obj %>%
  sleuth_fit(~Response,"full") %>%
  sleuth_fit(~1,"reduced") %>%
  sleuth_lrt("reduced","full")
```
```{r}
sleuth_diffex_lrt <- sleuth_results(s_obj,test="reduced:full",test_type="lrt")
sleuth_diffex_lrt %>% arrange(qval)
```

### Wald
```{r}
s_obj <- sleuth_wt(s_obj,"ResponseR")
sleuth_diffex_wald <- sleuth_results(s_obj,test="ResponseR",test_type="wald")
sleuth_diffex_lrt %>% arrange(qval)
```
```{r}
plot_volcano(s_obj,test_type="wt",test="ResponseR")
```

# DESeq (Kallisto)
```{r}
txi <- tximport(
  file.path(sleuth_samples$kallisto_path,"abundance.h5"),
  type = "kallisto",
  tx2gene = select(
    t2g,
    TXNAME = ensembl_transcript_id_version,
    GENEID = external_gene_name,
  ),
  ignoreAfterBar = TRUE,
)
deseq_kallisto_data <- DESeqDataSetFromTximport(txi,sleuth_samples,~Response)

```
## PCA
```{r}
deseq_kallisto_norm <- vst(deseq_kallisto_data,blind=TRUE,nsub = 20000)
plotPCA(deseq_kallisto_norm,intgroup="Response")
```
## Difex
```{r}
deseq_kallisto_data <- DESeq(deseq_kallisto_data)
deseq_kallisto_lrt <- results(deseq_kallisto_data,name="Response_R_vs_NR",tidy=TRUE)
deseq_kallisto_lrt %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>%
  select(row,log2FoldChange,pvalue,padj) %>%
  left_join(select(t2g,ensembl_gene_id,external_gene_name),by=c("row"="ensembl_gene_id"))
```

# DESeq (FeatureCounts)
```{r}
fc_mat <- read_tsv("../work/input/GSE126044_counts.txt")
 fc_mat <- fc_mat[!grepl("^\\d+\\-",fc_mat$X1),] %>%
   column_to_rownames("X1") %>%
   select(one_of(sleuth_samples$`Patient ID`))
deseq_fc <- DESeqDataSetFromMatrix(countData=fc_mat,colData=sleuth_samples,design=~Response)
```
## PCA
```{r}
deseq_fc_norm <- vst(deseq_fc)
plotPCA(deseq_fc_norm,intgroup="Response")
```

# Difex
```{r}
deseq_fc <- DESeq(deseq_fc)
deseq_fc_lrt <- results(deseq_fc,name="Response_R_vs_NR",tidy=TRUE)
deseq_fc_lrt %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>%
  select(row,log2FoldChange,pvalue,padj)
```

