---
title: "Comparisons of Differential Expression Tools on Different Data Sources"
output: html_notebook
---

# Setup
```{r}
library(biomaRt)
library(DESeq2)
library(sleuth)
library(tximport)
library(jsonlite)
library(data.table)
library(vsn)
library(tidyverse)
set.seed(42)
```

## Setup sample details
```{r}
# Load in samples and point to kallisto paths
samples <- read_tsv("../work/input/sample_manifest.tsv")
samples$kallisto_path <- file.path("..","work","intermediate","quant",samples$SRR)
```

## Transcript <-> Gene Mappings
```{r}
mart <- biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL",
  dataset = "hsapiens_gene_ensembl",
  host = 'grch37.ensembl.org')
gene_features <- biomaRt::getBM(attributes = c("ensembl_transcript_id_version", "ensembl_gene_id",
    "external_gene_name","transcript_length"), mart = mart)
t2g <- gene_features %>%
  select(ensembl_transcript_id_version,ensembl_gene_id,external_gene_name)
avg_tx_len <-gene_features %>%
  group_by(external_gene_name) %>%
  summarise(length=mean(transcript_length))
```

## QC Filtering
We'll filter high quality samples by percent reads mapped by Kallisto. We'll need to get hte total reads in files from the kallisto logs, and the total number of mapped reads by pre-loading all the counts using sleuth. 
```{r}
# Find total reads to compute percent mapped
samples$n_reads <- apply(samples,1,function(x)read_json(file.path(x["kallisto_path"],"run_info.json"))$n_processed)
```

```{r}
# Initial sleuth just to get kallisto counts
s_obj_counts <- sleuth_prep(
  sample_to_covariates = rename(samples,path=kallisto_path,sample=SRR),
  normalize=FALSE
)
```

```{r}
kallisto_est_mapped <- s_obj_counts$obs_raw %>%
  group_by(sample) %>%
  summarise(kallisto_mapped_reads=sum(est_counts))
samples <- samples %>% 
  left_join(kallisto_est_mapped,by=c("SRR"="sample")) %>%
  mutate(percent_mapped=kallisto_mapped_reads/n_reads) 
samples %>%
  ggplot(aes(x = percent_mapped)) +
    geom_histogram(bins = 10) +
    labs(title = "Distribution of % Mapped Reads", x = "% Mapped Reads")
```

```{r}
# Filter samples
samples_filt <-samples %>% 
  filter(
    !SRR %in% c("SRR8526726"),
    #percent_mapped > 0.8
    )
```



# Sleuth


```{r include=FALSE}
# Setup sleuth object
s_obj <- sleuth_prep(
  sample_to_covariates = rename(samples_filt,path=kallisto_path,sample=SRR),
  full_model = ~Response,
  transform_fun_counts=function(x)log2(x+1),
  target_mapping = rename(t2g,target_id=ensembl_transcript_id_version) %>%
      distinct(target_id,external_gene_name),
  aggregation_column = "external_gene_name",
  gene_mode=TRUE
)
```

## PCA 

### Sleuth default
```{r}
plot_pca(s_obj,units="scaled_reads_per_base",color_by = "Response",use_filtered = TRUE,text_labels = F) + geom_label(aes(label=sample),size=2)
```

### Standardized PCA
```{r}
sleuth_rna_norm <- s_obj$obs_norm_filt %>% 
  select(sample,target_id,scaled_reads_per_base) %>%
  spread(key=target_id,value=scaled_reads_per_base) %>%
  column_to_rownames("sample")
sleuth_std_pca <- prcomp((sleuth_rna_norm),center=T,scale.=T)
sleuth_std_pca$x[,1:2] %>%
  as.data.frame() %>%
  rownames_to_column("SRR") %>%
  left_join(samples,by="SRR") %>%
  ggplot(aes(x=PC1,y=PC2,color=Response,label=SRR)) +
    geom_point()  + 
   geom_label(size=2)
```

## Diffex

### LRT
```{r}
s_obj <- s_obj %>%
  sleuth_fit(~Response,"full") %>%
  sleuth_fit(~1,"reduced") %>%
  sleuth_lrt("reduced","full")
```
```{r}
sleuth_diffex_lrt <- sleuth_results(s_obj,test="reduced:full",test_type="lrt") %>% filter(!is.na(pval))
sleuth_diffex_lrt %>% arrange(qval)
```

### Wald
```{r}
s_obj <- sleuth_wt(s_obj,"ResponseR")
sleuth_diffex_wald <- sleuth_results(s_obj,test="ResponseR",test_type="wald")  %>% filter(!is.na(pval))
sleuth_diffex_wald %>% arrange(qval) 
```
```{r}
plot_volcano(s_obj,test_type="wt",test="ResponseR",sig_level=0.05) +
  labs(title="Volcano Plot of Sleuth Identified Differentially Expressed Genes",
       x="log2FC",
        color="Q < 0.05")
```

## QC
```{r}
meanSdPlot(log2(sleuth_to_matrix(s_obj,"obs_norm","scaled_reads_per_base") + 1),rank=FALSE,bins=100)
```

# DESeq (Kallisto)
```{r}
files <- file.path(samples_filt$kallisto_path,"abundance.h5")
names(files) <- samples_filt$`Patient ID`
txi <- tximport(
  file.path(samples_filt$kallisto_path,"abundance.h5"),
  type = "kallisto",
  tx2gene = select(
    t2g,
    TXNAME = ensembl_transcript_id_version,
    GENEID = external_gene_name
  ),
  ignoreAfterBar = TRUE,
)
rownames(samples_filt) <- samples_filt$`Patient ID`
deseq_kallisto_data <- DESeqDataSetFromTximport(txi,samples_filt,~Response)

```

## QC
```{r}
deseq_kallisto_norm <- vst(deseq_kallisto_data)
meanSdPlot(assay(deseq_kallisto_norm),ranks=FALSE,bins=100)
```


## PCA
```{r}
plotPCA(deseq_kallisto_norm,intgroup="Response",ntop=500000)
```
## Difex
```{r}
deseq_kallisto_data <- DESeq(deseq_kallisto_data)
deseq_kallisto_lrt <- results(deseq_kallisto_data,name="Response_R_vs_NR",tidy=TRUE)
deseq_kallisto_lrt %>% 
  #filter(padj < 0.05) %>%
  arrange(padj) %>%
  select(row,log2FoldChange,pvalue,padj) %>%
  left_join(select(t2g,ensembl_gene_id,external_gene_name),by=c("row"="ensembl_gene_id"))
```

# DESeq (FeatureCounts)
```{r}
fc_mat <- read_tsv("../work/input/GSE126044_counts.txt")
 fc_mat <- fc_mat[!grepl("^\\d+\\-",fc_mat$X1),] %>%
   column_to_rownames("X1") %>%
   select(one_of(samples_filt$`Patient ID`))
deseq_fc <- DESeqDataSetFromMatrix(countData=fc_mat,colData=samples_filt,design=~Response)
```

## QC
```{r}
deseq_fc_norm <- vst(deseq_fc)
meanSdPlot(assay(deseq_fc_norm),rank=FALSE,bins=100)
```


## PCA
```{r}
plotPCA(deseq_fc_norm,intgroup="Response",ntop=500000)
```

## Difex
```{r}
deseq_fc <- DESeq(deseq_fc)
deseq_fc_lrt <- results(deseq_fc,name="Response_R_vs_NR",tidy=TRUE)
deseq_fc_lrt %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>%
  select(row,log2FoldChange,pvalue,padj)
```

# Merged Analyses

## counts_to_rpm, pca
```{r}
tfc <- t(fc_mat)
# Convert Kallisto counts to rpm
kalread<-cbind(kallisto_mapped_reads=samples_filt$kallisto_mapped_reads, sleuth_rna_norm %>% rownames_to_column("sample"))
kalrpm<-setDT(kalread) [, lapply(.SD, function(x) (x * 1e6) / (kallisto_mapped_reads)),  by=.(kallisto_mapped_reads,sample)] %>% 
  column_to_rownames("sample")

# Convert star fc to rpm
#get df with est total reads for star
staread<-cbind(star_est_mapped=colSums(fc_mat), as.data.frame(tfc) %>% rownames_to_column("sample"))
starpm<-setDT(staread) [, lapply(.SD, function(x) (x * 1e6) / (star_est_mapped)),  by=.(star_est_mapped,sample)] %>%
  column_to_rownames("sample")

## scale pca with rpm
kal_rpm_pca <- prcomp(kalrpm[,-1],center=TRUE,scale.=TRUE)

starnorm_std <- scale(starpm[,-1])
#star counts include some zero, set scale/std apply to those !=0
star_rpm_pca <- prcomp(starnorm_std[ , which(apply(starnorm_std, 2, var) != 0)],center=TRUE)

## superimposed pca plots from the two sources
p1<-kal_rpm_pca$x[,1:2] %>% as.data.frame()
p2<-star_rpm_pca$x[,1:2] %>% as.data.frame()
ggplot()+geom_point(data=p1,aes(x=PC1,y=PC2, color='Kallisto')) + geom_point(data=p2,aes(x=PC1,y=PC2,color='STAR')) + scale_colour_manual(name="Dataset",values=c(Kallisto="blue", STAR="red"))
```


## Comparison of log2FC
```{r}
log2fc_combined <- sleuth_diffex_wald %>%
  filter(!is.na(b)) %>%
  select(gene=target_id,sleuth_log2FC=b,sleuth_q=qval) %>%
  left_join(select(deseq_kallisto_lrt,row,deseq_kal_log2FC=log2FoldChange,deseq_kal_q=padj),by=c("gene"="row")) %>%
  left_join(select(deseq_fc_lrt,row,deseq_fc_log2FC=log2FoldChange,deseq_fc_q=padj),by=c("gene"="row")) %>%
  filter(!is.na(deseq_fc_log2FC))
  
log2fc_combined %>%
  ggplot(aes(x=sleuth_log2FC,y=deseq_kal_log2FC)) +
    geom_point() +
    labs(title="Comparison of log2FC between Sleuth and DESeq Kallisto",
         subtitle=paste("R^2:",round(cor(log2fc_combined$sleuth_log2FC,log2fc_combined$deseq_kal_log2FC)^2,3)))
log2fc_combined %>%
  ggplot(aes(x=deseq_kal_log2FC,y=deseq_fc_log2FC)) +
    geom_point() +
    labs(title="Comparison of log2FC between DESeq From Kallisto and FC",
         subtitle=paste("R^2:",round(cor(log2fc_combined$deseq_kal_log2FC,log2fc_combined$deseq_fc_log2FC)^2,3)))
```

# Write out relevant files
```{r}
sleuth_diffex_lrt %>% write_tsv("sleuth_diffex_lrt.tsv")
sleuth_diffex_wald %>% write_tsv("sleuth_diffex_wald.tsv")

deseq_kallisto_lrt %>% write_tsv("deseq_kallisto_lrt.tsv")

deseq_fc_lrt %>% write_tsv("deseq_fc_lrt.tsv")
```


